File: java/1721_SwappingNodesInALinkedList.java
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        int idx = 0;
        ListNode curr = head, temp = curr;
        while (curr != null) {
            if (++idx == k)
                temp = curr;
            curr = curr.next;
        }
        
        int sz = idx;
        curr = head;
        for (int i = 0; i < sz - k; ++i) curr = curr.next;
        
        int currVal = curr.val;
        curr.val = temp.val;
        temp.val = currVal;
        
        return head;
    }
}

File: cpp/1721_SwappingNodesInALinkedList.cpp
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        if (!head) return head;
        
        int i = 1;
        ListNode* curr = head;
        ListNode* temp;
        while (curr) {
            if (i++ == k) {
                temp = curr;
            }
            curr = curr->next;
        }
        
        int size = i;
        i = 1;
        curr = head;
        while (curr) {
            if (size - k == i++) {
                int tmp = temp->val;
                temp->val = curr->val;
                curr->val = tmp;
            }
            curr = curr->next;
        }
        return head;
    }
};
